core/utils/codel.h

```diff
@@ -0,0 +1,301 @@
+#ifndef BESS_UTILS_CODEL_H_
+#define BESS_UTILS_CODEL_H_
+
+#include <cstring>
+#include <iostream>
+#include <string>
+
+#include "../kmod/llring.h"
+#include "../mem_alloc.h"
+#include "../packet.h"
+#include "time.h"
+
+#define DEFAULT_TARGET .05  // default delay target for codel
+#define DEFAULT_WINDOW .1   // default window size for codel
+#define DEFAULT_SLOTS 4096  // default number of slots in the codel queue
+
+/*
+  Codel(Controlled Delay Management) is an Queue controller based on this
+  article http://queue.acm.org/detail.cfm?id=2209336
+
+  It provides an active queue management to help prevent bufferbloat by dropping
+  queue entries at an increasing rate if the delay in the queue is above the
+  target queue delay. The equation used to calculate drop intervals is based on TCP
+  throughput response to drop probability.
+*/
+struct Codel {
```

Comment:
Josh, Panda and I discussed your reason for structuring it as close to llring as possible, and given that he also has concerns about it, I think the best bet is to write a parent class that is a generic queue wrapper class and have two child classes, one that is for llring and one that is for codel.  Melvin already wrote the llring wrapper, which you can find in the latest few commits here:
https://github.com/melvinw/bess/commits/work
So you can go take that, change it to inherit from a common parent class, and change codel to match the interface.